### PR TITLE
Support Multiple S3 and Docker Endpoints, and Offering Vault Integration alongside SSM

### PR DESCRIPTION
**Pull request by umang8223**
### Description
- Generalizing the GitHub-actions python scripts to enable them to run in environments outside of AWS.
- Why are we parameterizing the S3 region? The S3 region mentioned in our case differs from the region mentioned in the Python scripts, prompting the need for parameterization. This enhancement enables users to utilize different S3 regions as per their requirements.
- Why are we modifying the get_parameter_from_ssm function? The integration of Vault allows us to leverage its capabilities, enabling the execution of GitHub Actions on any cloud platform that supports Vault. It offers the flexibility to use Vault as an alternative to the standard SSM.
- Why are we modifying the login method?  This modification is necessary due to our utilization of a distinct Docker registry. It allows users to authenticate and access different Docker registries with their respective credentials.

### Changelog category (leave one):
- Build/Testing/Packaging Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Updated S3 region configuration in tests/ci/s3_helper.py to use the environment variable S3_REGION.
- Modified get_parameter_from_ssm function in tests/ci/get_robot_token.py to support fetching parameters from Vault if VAULT_URL is set.
- Enhanced pull request URL generation in tests/ci/clickhouse_helper.py by utilizing the environment variable GITHUB_REPOSITORY.
- Added Docker login credentials (DOCKER_USER and DOCKER_REPO) in tests/ci/env_helper.py for improved authentication.
- Updated default values and added the S3_REGION, S3_ENDPOINT and VAULT environment variable in tests/ci/env_helper.py.
- Improved image name construction in tests/ci/docker_images_check.py by including DOCKER_REPO.
- Enabled Docker login using DOCKER_USER and DOCKER_REPO in tests/ci/docker_images_check.py and tests/ci/docker_manifests_merge.py.
- Utilized S3_REGION environment variable for session creation in tests/ci/s3_helper.py.
- Updated the Docker image references in the Dockerfiles to utilize a parameterized Docker registry reference, ‘DOCKER_REPO’, with a default value of docker.io. This modification allows users to specify their desired Docker registry, providing greater flexibility and customization options.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by bkuschel** [comment]
@Felixoid @alexey-milovidov these changes will allow us to work and run the pipelines so that we can add support for other architectures and build variants. We'll be following up with similar changes for images and scripts. They should run, by default as they do now.

File: tests/ci/docker_images_check.py [hunk 1]
```diff
@@ -113,7 +119,7 @@ def get_changed_docker_images(
     for dockerfile_dir, image_description in images_dict.items():
         for f in files_changed:
             if f.startswith(dockerfile_dir):
-                name = image_description["name"]
+                name = DOCKER_REPO + "/" + image_description["name"]
```
Comment (Felixoid) [inline tests/ci/docker_images_check.py:122]: f-string looks more consistent here and below

Comment (Felixoid) [reply tests/ci/docker_images_check.py:122]: besides that, did you launch https://github.com/ClickHouse/ClickHouse/blob/master/.github/workflows/pull_request.yml#L48-L50 locally?

**Review by Felixoid — CHANGES_REQUESTED** [review]
Thanks for contributing!

There are some optional and some required changes in the PR

File: tests/ci/get_robot_token.py [hunk 2]
```diff
@@ -16,9 +19,19 @@ class Token:
 
 
 def get_parameter_from_ssm(name, decrypt=True, client=None):
-    if not client:
-        client = boto3.client("ssm", region_name="us-east-1")
-    return client.get_parameter(Name=name, WithDecryption=decrypt)["Parameter"]["Value"]
+    if VAULT_URL:
```
[the remark this answers is not in the document]
Comment (umang8223) [reply tests/ci/get_robot_token.py:21]: - Added import **hvac** inside the function.
- Added **hvac** in the **docker/test/style** Dockerfile.

File: tests/ci/get_robot_token.py [hunk 3]
```diff
@@ -28,18 +41,41 @@ def get_best_robot_token(token_prefix_env_name="github_robot_token_"):
...
-    )["Parameters"]
-    assert parameters
-
-    for token_name in [p["Name"] for p in parameters]:
-        value = get_parameter_from_ssm(token_name, True, client)
+
+    client = None
+    values = []
+
+    if VAULT_URL:
```
[the remark this answers is not in the document]
Comment (umang8223) [reply tests/ci/get_robot_token.py]: - Yes, certainly. I have made the necessary updates.

File: tests/ci/worker/prepare-ci-ami.sh [hunk 4]
```diff
@@ -90,7 +90,7 @@ systemctl restart docker
 sudo -u ubuntu docker buildx version
 sudo -u ubuntu docker buildx create --use --name default-builder
 
-pip install boto3 pygithub requests urllib3 unidiff dohq-artifactory
+pip install boto3 pygithub requests urllib3 unidiff dohq-artifactory hvac
```
[the remark this answers is not in the document]
Comment (umang8223) [reply tests/ci/worker/prepare-ci-ami.sh]: - A condition has been added to install hvac only when running in our environment.

File: tests/ci/env_helper.py [hunk 5]
```diff
@@ -23,8 +25,14 @@
 REPORTS_PATH = os.getenv("REPORTS_PATH", p.abspath(p.join(module_dir, "./reports")))
 REPO_COPY = os.getenv("REPO_COPY", git_root)
 RUNNER_TEMP = os.getenv("RUNNER_TEMP", p.abspath(p.join(module_dir, "./tmp")))
+S3_REGION = os.getenv("S3_REGION", "us-east-1")
 S3_BUILDS_BUCKET = os.getenv("S3_BUILDS_BUCKET", "clickhouse-builds")
 S3_TEST_REPORTS_BUCKET = os.getenv("S3_TEST_REPORTS_BUCKET", "clickhouse-test-reports")
+S3_ENDPOINT = os.getenv("S3_ENDPOINT", "https://s3.amazonaws.com")
```
[the remark this answers is not in the document]
Comment (umang8223) [reply tests/ci/env_helper.py:31]: - Yes thats right.

File: tests/ci/docker_images_check.py (hunk 1 above)
Comment (umang8223) [reply tests/ci/docker_images_check.py]: I have updated it to an f-string.

File: docker/test/style/Dockerfile [hunk 6]
```diff
@@ -18,7 +18,11 @@ RUN apt-get update && env DEBIAN_FRONTEND=noninteractive apt-get install --yes \
     python3-pip \
     shellcheck \
     yamllint \
-    && pip3 install black==23.1.0 boto3 codespell==2.2.1 dohq-artifactory mypy PyGithub unidiff pylint==2.6.2 \
+    && if [ "$DOCKER_REPO" = "docker.io" ]; then \
+         pip3 install black==23.1.0 boto3 codespell==2.2.1 dohq-artifactory mypy PyGithub unidiff pylint==2.6.2; \
+       else \
```
Comment (Felixoid) [inline docker/test/style/Dockerfile:23]:
```suggestion
       else \
       # the `hvac` is an optional dependency for Hashicorp Vault as a parameters store
```

File: tests/ci/docker_images_check.py [hunk 7]
```diff
@@ -94,7 +100,7 @@ def get_images_dict(repo_path: str, image_file_path: str) -> ImagesDict:
 
 
 def get_changed_docker_images(
-    pr_info: PRInfo, images_dict: ImagesDict
+    pr_info: PRInfo, images_dict: ImagesDict, DOCKER_REPO
```
Comment (Felixoid) [inline tests/ci/docker_images_check.py:103]: mypy will fail here, parameter type is required

File: tests/ci/env_helper.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Felixoid) [reply tests/ci/env_helper.py:31]: Please, remove `S3_ENDPOINT` in favor of the existing `S3_URL`. The parameters are also ordered alphabetically, so `VAULT_*` should be at the end.

File: tests/ci/s3_helper.py [hunk 8]
```diff
@@ -47,7 +48,7 @@ class S3Helper:
 
     def __init__(self):
         config = botocore.config.Config(max_pool_connections=self.max_pool_size)
-        self.session = boto3.session.Session(region_name="us-east-1")
+        self.session = boto3.session.Session(region_name=S3_REGION)
```
Comment (Felixoid) [inline tests/ci/s3_helper.py:51]: `us-east-1` is seen in many other places, as well as in new once. Let's use `S3_REGION` in get_robot_token.py and jepsen_check.py too

File: tests/ci/worker/prepare-ci-ami.sh [hunk 9]
```diff
@@ -90,7 +90,11 @@ systemctl restart docker
 sudo -u ubuntu docker buildx version
 sudo -u ubuntu docker buildx create --use --name default-builder
 
-pip install boto3 pygithub requests urllib3 unidiff dohq-artifactory
+if [[ -n "${VAULT_URL}" ]]; then
```
Comment (Felixoid) [inline tests/ci/worker/prepare-ci-ami.sh:93]: Where it would be set?

**Comment by Felixoid** [comment]
Let's check how it will work in our infrastructure

**Comment by robot-ch-test-poll1** [comment]
This is an automated comment for commit 9bcfc18bc636f3d3f66e9ff6f1a67ad0972d5578 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/50630/9bcfc18bc636f3d3f66e9ff6f1a67ad0972d5578/ci_running.html)
The overall status of the commit is 🔴 failure

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>Style Check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td>🔴 failure</td></tr>
</table>

**Comment by Felixoid** [comment]
looks very bad `¯\_(ツ)_/¯`

**Comment by umang8223** [comment]
> looks very bad `¯\_(ツ)_/¯`

Sorry for the delay, I will look for the root cause and fix it. Thanks

**Comment by alexey-milovidov** [comment]
Please reopen when ready.

**Pull request closed without merging**